{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "AFIpbimuEBkc"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "source": [
        "##############################################################################\n",
        "# PPRInt2.0 is developed for predicting the RNA-interacting residues in protein  #\n",
        "# using primary structure. It is developed by Prof G. P. S. Raghava's group.  #\n",
        "# Please cite: PPRInt2.0\t\t\t\t\t\t      #\n",
        "# ############################################################################\n",
        "import os\n",
        "import pickle\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import sys\n",
        "import csv\n",
        "import re\n",
        "import glob\n",
        "import time\n",
        "from time import sleep\n",
        "#from tqdm import tqdm\n",
        "import argparse\n",
        "import warnings\n",
        "from argparse import RawTextHelpFormatter\n",
        "import tensorflow as tf\n",
        "from keras.models import load_model\n",
        "import uuid\n",
        "warnings.filterwarnings('ignore')\n",
        "\n",
        "\n",
        "parser = argparse.ArgumentParser(description='Please provide following arguments',formatter_class=RawTextHelpFormatter)\n",
        "## Read Arguments from command\n",
        "parser.add_argument(\"-i\", \"--input\", type=str, required=True, help=\"Input: Protein sequences in FASTA format\")\n",
        "parser.add_argument(\"-o\", \"--output\",type=str, help=\"Output: File for saving results by default outfile.csv\")\n",
        "parser.add_argument(\"-j\", \"--job\",type=int, choices = [1,2,3,4], help=\"Feature Type:\\n1: Amino Acid Binary (AAB) profile\\n2: Physico-chemical based binary (PCB) profile\\n3: Position-Specific Scoring Matrix (PSSM) profile\\n4: Hybrid (AAB+PCB+PSSM), by default 3 (PSSM)\")\n",
        "parser.add_argument(\"-t\",\"--threshold\", type=float, help=\"Threshold: Value between 0 to 1 by default 0.5\")\n",
        "parser.add_argument(\"-p\", \"--path\",type=str, help=\"Path: Please provide the path of python which has all libraries installed\")\n",
        "args = parser.parse_args()\n",
        "\n",
        "def readseq(file):\n",
        "    with open(file) as f:\n",
        "        records = f.read()\n",
        "    records = records.split('>')[1:]\n",
        "    seqid = []\n",
        "    seq = []\n",
        "    for fasta in records:\n",
        "        array = fasta.split('\\n')\n",
        "        name, sequence = array[0].split()[0], re.sub('[^ARNDCQEGHILKMFPSTWYV-]', '', ''.join(array[1:]).upper())\n",
        "        seqid.append(name)\n",
        "        seq.append(sequence)\n",
        "    if len(seqid) == 0:\n",
        "        f=open(file,\"r\")\n",
        "        data1 = f.readlines()\n",
        "        for each in data1:\n",
        "            seq.append(each.replace('\\n',''))\n",
        "        for i in range (1,len(seq)+1):\n",
        "            seqid.append(\"Seq_\"+str(i))\n",
        "    for i in seq:\n",
        "        if 'B' in i:\n",
        "            print('\\nError: The input sequences contain non-natural amino acids. Kindly check the sequence.\\n')\n",
        "            sys.exit()\n",
        "        if 'J' in i:\n",
        "            print('\\nError: The input sequences contain non-natural amino acids. Kindly check the sequence.\\n')\n",
        "            sys.exit()\n",
        "        if 'O' in i:\n",
        "            print('\\nError: The input sequences contain non-natural amino acids. Kindly check the sequence.\\n')\n",
        "            sys.exit()\n",
        "        if 'U' in i:\n",
        "            print('\\nError: The input sequences contain non-natural amino acids. Kindly check the sequence.\\n')\n",
        "            sys.exit()\n",
        "        if 'Z' in i:\n",
        "            print('\\nError: The input sequences contain non-natural amino acids. Kindly check the sequence.\\n')\n",
        "            sys.exit()\n",
        "        if 'X' in i:\n",
        "            print('\\nError: The input sequences contain non-natural amino acids. Kindly check the sequence.\\n')\n",
        "            sys.exit()\n",
        "    df4 = pd.DataFrame(seq)\n",
        "    df5 = pd.DataFrame(seqid)\n",
        "    return df4,df5\n",
        "def aab(file,out):\n",
        "    std = list('ACDEFGHIKLMNPQRSTVWYX')\n",
        "    df = pd.read_csv(file, header = None)\n",
        "    uu = []\n",
        "    for ss in df[0]:\n",
        "         uu.append(len(ss))\n",
        "    zz = df.iloc[:,0]\n",
        "    f = open(out, mode='w')\n",
        "    sys.stdout = f\n",
        "    A=('1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    C=('0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    D=('0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    E=('0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    F=('0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    G=('0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    H=('0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    I=('0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    K=('0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    L=('0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    M=('0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0')\n",
        "    N=('0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0')\n",
        "    P=('0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0')\n",
        "    Q=('0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0')\n",
        "    R=('0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0')\n",
        "    S=('0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0')\n",
        "    T=('0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0')\n",
        "    V=('0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0')\n",
        "    W=('0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0')\n",
        "    Y=('0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0')\n",
        "    X=('0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1')\n",
        "    for mm in range (1,358):\n",
        "        print(\"F_\"+str(mm),end=',')\n",
        "    print(\"\")\n",
        "    for i in range(0,len(zz)):\n",
        "        for j in zz[i]:\n",
        "            if j == \"A\":\n",
        "                print(''.join(A), end = ',')\n",
        "            if j == \"C\":\n",
        "                print(''.join(C), end = ',')\n",
        "            if j == \"D\":\n",
        "                print(''.join(D), end = ',')\n",
        "            if j == \"E\":\n",
        "                print(''.join(E), end = ',')\n",
        "            if j == \"F\":\n",
        "                print(''.join(F), end = ',')\n",
        "            if j == \"G\":\n",
        "                print(''.join(G), end = ',')\n",
        "            if j == \"H\":\n",
        "                print(''.join(H), end = ',')\n",
        "            if j == \"I\":\n",
        "                print(''.join(I), end = ',')\n",
        "            if j == \"K\":\n",
        "                print(''.join(K), end = ',')\n",
        "            if j == \"L\":\n",
        "                print(''.join(L), end = ',')\n",
        "            if j == \"M\":\n",
        "                print(''.join(M), end = ',')\n",
        "            if j == \"N\":\n",
        "                print(''.join(N), end = ',')\n",
        "            if j == \"P\":\n",
        "                print(''.join(P), end = ',')\n",
        "            if j == \"Q\":\n",
        "                print(''.join(Q), end = ',')\n",
        "            if j == \"R\":\n",
        "                print(''.join(R), end = ',')\n",
        "            if j == \"S\":\n",
        "                print(''.join(S), end = ',')\n",
        "            if j == \"T\":\n",
        "                print(''.join(T), end = ',')\n",
        "            if j == \"V\":\n",
        "                print(''.join(V), end = ',')\n",
        "            if j == \"W\":\n",
        "                print(''.join(W), end = ',')\n",
        "            if j == \"Y\":\n",
        "                print(''.join(Y), end = ',')\n",
        "            if j == \"X\":\n",
        "                print(''.join(X), end = ',')\n",
        "        print(\"\")\n",
        "    f.truncate()\n",
        "def pcp_bin(file,out):\n",
        "    std = list('ACDEFGHIKLMNPQRSTVWYX')\n",
        "    df = pd.read_csv(file, header = None)\n",
        "    uu = []\n",
        "    for ss in df[0]:\n",
        "         uu.append(len(ss))\n",
        "    zz = df.iloc[:,0]\n",
        "    f = open(out, mode='w')\n",
        "    sys.stdout = f\n",
        "    A=('0,0,1,0,1,1,0,0,0,0,1,1,0,0,0,0,1,0,0,1,0,0,1,1,0')\n",
        "    C=('0,0,1,1,0,0,0,0,0,0,1,1,0,0,0,1,0,1,0,1,0,0,1,1,0')\n",
        "    D=('0,1,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,1,0,0,1,0')\n",
        "    E=('0,1,0,0,0,0,0,0,1,0,0,0,0,1,0,0,1,0,0,0,1,0,0,0,1')\n",
        "    F=('0,0,1,0,1,0,0,1,0,0,1,1,0,0,0,0,0,1,0,1,0,0,0,0,1')\n",
        "    G=('0,0,1,0,1,1,0,0,0,0,1,0,0,1,0,0,0,0,1,1,0,0,1,1,0')\n",
        "    H=('1,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,1,0,0,0,0,1,0,0,1')\n",
        "    I=('0,0,1,0,1,1,0,0,0,0,1,1,0,0,0,0,0,1,0,1,0,0,0,0,1')\n",
        "    K=('1,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1')\n",
        "    L=('0,0,1,0,1,1,0,0,0,0,1,1,0,0,0,0,1,0,0,1,0,0,0,0,1')\n",
        "    M=('0,0,1,0,1,0,0,0,0,0,1,1,0,0,0,1,1,0,0,0,0,1,0,0,1')\n",
        "    N=('0,0,1,0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,1,0,1,0,0,1,0')\n",
        "    P=('0,0,1,0,1,1,1,0,0,0,1,1,1,0,0,0,0,0,1,0,1,1,0,1,0')\n",
        "    Q=('0,0,1,1,0,0,0,0,0,0,1,0,0,1,0,0,1,0,0,0,1,0,0,0,1')\n",
        "    R=('1,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,1,0,0,0,0,0,0,0,1')\n",
        "    S=('0,0,1,1,0,0,0,0,0,0,1,0,0,1,1,0,0,0,1,0,0,1,1,1,0')\n",
        "    T=('0,0,1,1,0,0,0,0,0,0,1,1,0,1,1,0,0,1,0,0,0,1,0,1,0')\n",
        "    V=('0,0,1,0,1,1,0,0,0,0,1,1,0,0,0,0,0,1,0,1,0,0,0,1,0')\n",
        "    W=('0,0,1,0,1,0,0,1,0,0,1,1,0,0,0,0,0,1,0,1,0,0,0,0,1')\n",
        "    Y=('0,0,1,1,0,0,0,1,0,0,1,0,0,0,0,0,0,1,0,0,0,1,0,0,1')\n",
        "    X=('0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0')\n",
        "    for mm in range (358,783):\n",
        "        print(\"F_\"+str(mm),end=',')\n",
        "    print(\"\")\n",
        "    for i in range(0,len(zz)):\n",
        "        for j in zz[i]:\n",
        "            if j == \"A\":\n",
        "                print(''.join(A), end = ',')\n",
        "            if j == \"C\":\n",
        "                print(''.join(C), end = ',')\n",
        "            if j == \"D\":\n",
        "                print(''.join(D), end = ',')\n",
        "            if j == \"E\":\n",
        "                print(''.join(E), end = ',')\n",
        "            if j == \"F\":\n",
        "                print(''.join(F), end = ',')\n",
        "            if j == \"G\":\n",
        "                print(''.join(G), end = ',')\n",
        "            if j == \"H\":\n",
        "                print(''.join(H), end = ',')\n",
        "            if j == \"I\":\n",
        "                print(''.join(I), end = ',')\n",
        "            if j == \"K\":\n",
        "                print(''.join(K), end = ',')\n",
        "            if j == \"L\":\n",
        "                print(''.join(L), end = ',')\n",
        "            if j == \"M\":\n",
        "                print(''.join(M), end = ',')\n",
        "            if j == \"N\":\n",
        "                print(''.join(N), end = ',')\n",
        "            if j == \"P\":\n",
        "                print(''.join(P), end = ',')\n",
        "            if j == \"Q\":\n",
        "                print(''.join(Q), end = ',')\n",
        "            if j == \"R\":\n",
        "                print(''.join(R), end = ',')\n",
        "            if j == \"S\":\n",
        "                print(''.join(S), end = ',')\n",
        "            if j == \"T\":\n",
        "                print(''.join(T), end = ',')\n",
        "            if j == \"V\":\n",
        "                print(''.join(V), end = ',')\n",
        "            if j == \"W\":\n",
        "                print(''.join(W), end = ',')\n",
        "            if j == \"Y\":\n",
        "                print(''.join(Y), end = ',')\n",
        "            if j == \"X\":\n",
        "                print(''.join(X), end = ',')\n",
        "        print(\"\")\n",
        "    f.truncate()\n",
        "def pssm_1(x):\n",
        "        if type(x) is str:\n",
        "            return x\n",
        "        elif x <= -32768:\n",
        "             return 0\n",
        "        elif x:\n",
        "            return (1/(1+(2.7182)**(-x)))\n",
        "        else:\n",
        "            return\n",
        "def pssm_2(file):\n",
        "    pd.options.display.float_format = '{:.2e}'.format\n",
        "    df=file\n",
        "    df1 = df\n",
        "    df2 = df1.applymap(pssm_1)\n",
        "    return df2\n",
        "def pssm(inputfile,outputfile):\n",
        "    if os.path.exists('envfile'):\n",
        "        with open('envfile', 'r') as file:\n",
        "            data = file.readlines()\n",
        "        output = []\n",
        "        for line in data:\n",
        "            if not \"#\" in line:\n",
        "                output.append(line)\n",
        "        if len(output)==3:\n",
        "            paths = []\n",
        "            for i in range (0,len(output)):\n",
        "                paths.append(output[i].split(':')[1].replace('\\n',''))\n",
        "            blastpgp = paths[0]\n",
        "            blastdb = paths[1]\n",
        "            makemat = paths[2]\n",
        "        if os.path.isfile(blastpgp) and os.access(blastpgp, os.R_OK):\n",
        "            print('The provided directory for blastpgp is correct and readable.')\n",
        "        else:\n",
        "            print(\"########################################################################################################################\")\n",
        "            print(\"Error: Either 'blastbgp' file is missing from the provided directory in the 'envfile', or not readable. Kindly check.\", file=sys.stderr)\n",
        "            print(\"########################################################################################################################\")\n",
        "            sys.exit()\n",
        "        if os.path.isfile(makemat) and os.access(makemat, os.R_OK):\n",
        "            print('The provided directory for makemat is correct and readable.')\n",
        "        else:\n",
        "            print(\"########################################################################################################################\")\n",
        "            print(\"Error: Either 'makemat' file is missing from the provided directory in the 'envfile', or not readable. Kindly check.\", file=sys.stderr)\n",
        "            print(\"########################################################################################################################\")\n",
        "            sys.exit()\n",
        "        if (glob.glob(blastdb+\".pin\")) and (glob.glob(blastdb+\".psq\")) and (glob.glob(blastdb+\".phr\")):\n",
        "            print('The provided directory for blast database is correct and readable.')\n",
        "        else:\n",
        "            dbfiles = blastdb.split('/')[-1]\n",
        "            print(\"##############################################################################################################################################################################################\")\n",
        "            print(\"Error: Either the files for BLAST database are missing from the provided directory in the 'envfile', or not readable. Please provide the files with extension of\", dbfiles+\".pin,\", dbfiles+\".phr,\", dbfiles+\".psq in the provided directory.\", \"Kindly check.\", file=sys.stderr)\n",
        "            print(\"##############################################################################################################################################################################################\")\n",
        "            sys.exit()\n",
        "\n",
        "    else:\n",
        "        print(\"####################################################################################\")\n",
        "        print(\"Error: Please provide the '{}', which comprises paths for BLASTPGP and MAKEMAT\".format('envfile'), file=sys.stderr)\n",
        "        print(\"####################################################################################\")\n",
        "        sys.exit()\n",
        "    ss = []\n",
        "    file_list = []\n",
        "    df1,dff12 = readseq(inputfile)\n",
        "    aa = []\n",
        "    for i in df1[0]:\n",
        "        aa.append(len(i))\n",
        "    for i in range(0,len(df1)):\n",
        "        ss.append('>seq_'+str(i+1))\n",
        "    df1['seq'] = ss\n",
        "    df1 = df1[['seq',0]]\n",
        "    for ii in range(0,len(df1)):\n",
        "        name_file = df1['seq'][ii].replace('>','')+'.fasta'\n",
        "        file_out = df1['seq'][ii].replace('>','')+'.pssmout'\n",
        "        file_list.append(df1['seq'][ii].replace('>','')+'.mtx')\n",
        "        df1.iloc[ii].to_csv(name_file, index=None, header=False, sep='\\n')\n",
        "        filename, file_extension = os.path.splitext(name_file)\n",
        "        filename_o, file_extension_o = os.path.splitext(file_out)\n",
        "        S =  blastpgp + ' -d ' +  blastdb + ' -i ' + name_file + ' -j 3 -C ' + file_out\n",
        "        os.system(S)\n",
        "        os.rename(file_out, filename_o + '.chk')\n",
        "        outputfile1 = filename_o + \".chk\"\n",
        "        temp1 =filename_o + \".sn\"\n",
        "        C2 = 'echo {} > {}'.format(name_file,temp1)\n",
        "        os.system(C2)\n",
        "        temp2 = filename_o + \".pn\"\n",
        "        C1 = 'echo {} > {}'.format(outputfile1,temp2)\n",
        "        os.system(C1)\n",
        "        P = makemat + ' -P ' + filename_o\n",
        "        os.system(P)\n",
        "    dir = '.'\n",
        "    tt = []\n",
        "    for i in file_list:\n",
        "        ss = []\n",
        "        fp = open(i)\n",
        "        all_line = fp.readlines()\n",
        "        ss.append(all_line[14:])\n",
        "        uu = []\n",
        "        for j in all_line[14:]:\n",
        "            uu.append(j.replace('\\n','').replace('  ',','))\n",
        "        np.savetxt(i+'_temp',uu, fmt=\"%s\")\n",
        "        df11 = pd.read_csv(i+'_temp', header=None, sep=\",\")\n",
        "        col = [1,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,22,23]\n",
        "        df13 = df11.iloc[:,col].reset_index(drop=True)\n",
        "        df12 = pssm_2(df13)\n",
        "        for i in range(0,len(df12)):\n",
        "            tt.extend(df12.loc[i])\n",
        "    bb = []\n",
        "    cc = 0\n",
        "    for i in range(0,len(aa)):\n",
        "        bb.append(tt[cc:cc+21*(aa[i])])\n",
        "        cc = cc+21*(aa[i])\n",
        "    df123 = pd.DataFrame(bb)\n",
        "    df456 = df123.fillna('NA')\n",
        "    df456.to_csv(outputfile,index=None,header=False,float_format='%.2e')\n",
        "    allfiles = os.listdir(dir)\n",
        "    for item in allfiles :\n",
        "        if item.endswith(\".aux\") or item.endswith(\".sn\") or item.endswith(\".pn\") or item.endswith(\".mn\") or item.endswith(\".mtx_temp\") or item.endswith(\".mtx\") or item.endswith(\".chk\") or item.endswith(\".fasta\") or item.endswith(\".readseq\"):\n",
        "            os.remove(os.path.join(dir, item))\n",
        "def pssm2pat(seqfile,infile,nom):\n",
        "    std = list('ACDEFGHIKLMNPQRSTVWYX')\n",
        "    df = seqfile\n",
        "    a = df[0].values\n",
        "    pp = []\n",
        "    tt = []\n",
        "    yy = []\n",
        "    df1 = infile\n",
        "    for i in range(0,len(df1)):\n",
        "        ss = df1.values.tolist()[i][:(len(a[i])*21)]\n",
        "        uu = [0]*21*(int((nom-1)/2))\n",
        "        mm = [0]*21*(int((nom-1)/2))\n",
        "        uu.extend(ss)\n",
        "        uu.extend(mm)\n",
        "        pp.append(uu)\n",
        "    for i in range(0,len(df1)):\n",
        "        for j in range(0,(len(pp[i])-(int(nom)*21))+1,21):\n",
        "            tt.append(pp[i][j:j+(21*int(nom))])\n",
        "    for i in range(1,nom+1):\n",
        "        for j in std:\n",
        "            yy.append(j+str(i))\n",
        "    df2 = pd.DataFrame(tt)\n",
        "    df2.columns = yy\n",
        "    return df2\n",
        "def seq2motif(data_1):\n",
        "    df = data_1\n",
        "    cc = []\n",
        "    for i in range(0,len(df)):\n",
        "        ss = 'XXXXXXXX'+df[0][i]+'XXXXXXXX'\n",
        "        for j in range(0,len(ss)):\n",
        "            if len(ss[j:j+17])==17:\n",
        "                cc.append(ss[j:j+17])\n",
        "    df6 = pd.DataFrame(cc)\n",
        "    return df6\n",
        "def matrix_gen(data_train):\n",
        "    matrix_input=np.array(data_train)\n",
        "    matrix_input.shape\n",
        "    matrix_3d_form_input= np.zeros([len(data_train),1,len(data_train.columns)], dtype=float)\n",
        "    matrix_3d_form_input.shape\n",
        "    for i in range(matrix_input.shape[0]):\n",
        "        row=matrix_input[i]\n",
        "        matrix_3d_form_input[i]=np.array(row)\n",
        "    return matrix_3d_form_input\n",
        "def pred(data_1,model_1):\n",
        "    X1 = matrix_gen(data_1)\n",
        "    model = load_model(model_1)\n",
        "    train_predictions_weighted = model.predict(X1,batch_size=64)\n",
        "    df5 = pd.DataFrame(train_predictions_weighted)\n",
        "    return df5\n",
        "def pred_hy(data_1,data_2,data_3,model_1):\n",
        "    X1 = matrix_gen(data_1)\n",
        "    X2 = matrix_gen(data_2)\n",
        "    X3 = matrix_gen(data_3)\n",
        "    model = load_model(model_1)\n",
        "    train_predictions_weighted = model.predict([X1,X2,X3],batch_size=64)\n",
        "    df5 = pd.DataFrame(train_predictions_weighted)\n",
        "    return df5\n",
        "def result_maker(data_1,data_2,data_3,data_4,num):\n",
        "    df = data_1\n",
        "    df1 = data_2\n",
        "    df2 = data_3\n",
        "    cc = []\n",
        "    dd = []\n",
        "    uu = []\n",
        "    mm = []\n",
        "    count = 0\n",
        "    for i in range(0,len(df1)):\n",
        "        if df1[0][i] >= num:\n",
        "            cc.append('+')\n",
        "        else:\n",
        "            cc.append('-')\n",
        "    df1['pre'] = cc\n",
        "    for i in range(0,len(df)):\n",
        "        dd.append(len(df[0][i]))\n",
        "    for i in dd:\n",
        "        ss = df1['pre'][count:count+i].values\n",
        "        uu.append(''.join(ss))\n",
        "    for i in range(0,len(df2)):\n",
        "        mm.append('>'+df2[0][i])\n",
        "    df3 = pd.DataFrame()\n",
        "    df3['ID'] = mm\n",
        "    df3['Seq'] = df[0]\n",
        "    df3['Pred'] = uu\n",
        "    df3.to_csv(data_4,index=None, header=False, sep=\"\\n\")\n",
        "('##############################################################################')\n",
        "print('# This program PPRInt2.0 is developed for predicting RNA-interacting resdiues  #')\n",
        "print('# in protein sequence, developed by Prof G. P. S. Raghava group. #')\n",
        "print('# ############################################################################')\n",
        "\n",
        "\n",
        "#variables to hold input and output files\n",
        "inFile= args.input\n",
        "\n",
        "# Output file\n",
        "\n",
        "if args.output == None:\n",
        "     outFile= \"outfile.csv\"\n",
        "else:\n",
        "     outFile = args.output\n",
        "\n",
        "if args.threshold == None:\n",
        "        threshold = 0.5\n",
        "else:\n",
        "        threshold= float(args.threshold)\n",
        "\n",
        "# Job Type\n",
        "if args.job == None:\n",
        "        Job = int(3)\n",
        "else:\n",
        "        Job = int(args.job)\n",
        "\n",
        "# Python Path\n",
        "if args.path == None:\n",
        "        ppath = 'python3'\n",
        "else:\n",
        "        ppath = args.path\n",
        "\n",
        "\n",
        "print('Summary of Parameters:')\n",
        "print('Input File: ',inFile,'; Threshold: ', threshold,'; Job Type: ',Job)\n",
        "print('Output File: ',outFile,)\n",
        "\n",
        "path_test=inFile\n",
        "path_output=outFile\n",
        "\n",
        "\"\"\"Calling composition py files\"\"\"\n",
        "#======================= Prediction Module start from here =====================\n",
        "file_gen = str(uuid.uuid4())\n",
        "s70,s71 = readseq(path_test)\n",
        "s71.to_csv('seqids.csv',index=None,header=None)\n",
        "if Job == 1:\n",
        "    print('==== Initiating the process of Prediction using amino acid binaray profiles as input features: please wait ...')\n",
        "    dff = seq2motif(s70)\n",
        "    dff.to_csv(file_gen, index=None, header=False)\n",
        "    aab(file_gen,'temp.bin')\n",
        "    df_aab = pd.read_csv('temp.bin')\n",
        "    df_aab_1 = df_aab.iloc[:,:-1]\n",
        "    res = pred(df_aab_1,'prog/AAB_model.h5')\n",
        "    result_maker(s70,res,s71,path_output,threshold)\n",
        "    os.remove('temp.bin')\n",
        "    os.remove('seqids.csv')\n",
        "    os.remove(file_gen)\n",
        "if Job ==2:\n",
        "    print('==== Initiating the process of Prediction using physico-chemical properties based  binaray profiles as input features: please wait ...')\n",
        "    dff = seq2motif(s70)\n",
        "    dff.to_csv(file_gen, index=None, header=False)\n",
        "    pcp_bin(file_gen,'temp.pcb')\n",
        "    df_pcb = pd.read_csv('temp.pcb')\n",
        "    df_pcb_1 = df_pcb.iloc[:,:-1]\n",
        "    res = pred(df_pcb_1,'prog/PCB_model.h5')\n",
        "    result_maker(s70,res,s71,path_output,threshold)\n",
        "    os.remove('temp.pcb')\n",
        "    os.remove('seqids.csv')\n",
        "    os.remove(file_gen)\n",
        "if Job ==3:\n",
        "    print('==== Initiating the process of Prediction using position specific scoring matrix profiles as input features: please wait ...')\n",
        "    pssm(path_test,'temp.pssm')\n",
        "    df_pssm = pd.read_csv('temp.pssm', header=None)\n",
        "    df_pssm_1 = pssm2pat(s70,df_pssm,17)\n",
        "    res = pred(df_pssm_1,'prog/PSSM_model.h5')\n",
        "    result_maker(s70,res,s71,path_output,threshold)\n",
        "    os.remove('temp.pssm')\n",
        "    os.remove('seqids.csv')\n",
        "    os.remove(file_gen)\n",
        "if Job==4:\n",
        "    print('==== Initiating the process of Prediction using hybrid of all features (AAB+PCB+PSSM)s as input features: please wait ...')\n",
        "    dff = seq2motif(s70)\n",
        "    dff.to_csv(file_gen, index=None, header=False)\n",
        "    aab(file_gen,'temp.bin')\n",
        "    pcp_bin(file_gen,'temp.pcb')\n",
        "    pssm(path_test,'temp.pssm')\n",
        "    df_aab = pd.read_csv('temp.bin')\n",
        "    df_pcb = pd.read_csv('temp.pcb')\n",
        "    df_pssm = pd.read_csv('temp.pssm', header=None)\n",
        "    df_aab_1 = df_aab.iloc[:,:-1]\n",
        "    df_pcb_1 = df_pcb.iloc[:,:-1]\n",
        "    df_pssm_1 = pssm2pat(s70,df_pssm,17)\n",
        "    res = pred_hy(df_aab_1,df_pcb_1,df_pssm_1,'prog/Hybrid_model.h5')\n",
        "    result_maker(s70,res,s71,path_output,threshold)\n",
        "    os.remove('temp.bin')\n",
        "    os.remove('temp.pcb')\n",
        "    os.remove('temp.pssm')\n",
        "    os.remove('seqids.csv')\n",
        "    os.remove(file_gen)\n",
        "\n",
        "print('\\n======= Thanks for using PPRInt2.0. Your results are stored in file :',outFile,' =====\\n\\n')\n",
        "print('Please cite: PPRInt2.0\\n')\n",
        "print(\"Et voila. The process is done!!!!!\")"
      ],
      "metadata": {
        "id": "Ns8b50nlEEBL"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}